cmd/lotus-chainwatch/storage.go

```diff
@@ -318,7 +326,47 @@ create table if not exists miner_sector_events
     
 	constraint miner_sector_events_pk
 		primary key (sector_id, event, miner_id, state_root)
-)
+);
+
+create table if not exists market_deal_proposals
+(
+    deal_id bigint not null,
+    
+    state_root text not null,
+    
+    piece_cid text not null,
+    padded_piece_size bigint not null,
+    unpadded_piece_size bigint not null,
+    verified_deal bool not null,
+    
+    client_id text not null,
+    provider_id text not null,
+    
+    start_epoch bigint not null,
+    end_epoch bigint not null,
+    storage_price_per_epoch text not null,
```

Comment:
Can a deal be slashed more than once? If not, can we include the slashed epoch in the schema here?

Reply:
Based on [this code](https://github.com/filecoin-project/specs-actors/blob/master/actors/builtin/market/market_state.go#L102) I would _think_ not.

We can get the slashed epoch by joining with the `market_deal_states` table, is there a reason we want the value here too?

Reply:
We're balancing the cost of JOINs against the ease of repeated, denormalized data. Given these tables are going to get VERY large, I err away from the JOIN cost. This also matches the pattern of the sector_info where we're surfacing the latest state out of convenience with the understanding that it's canonical representation (and any state transitions) are elsewhere.

Reply:
There will also be opportunities to tune these schemas as well.... I'm making some educated guesses which we can revisit once we have more experience.